### PR TITLE
fix: Fix cancellation (including deadlines) for REGAPIC streaming

### LINKED ISSUE #648: Indicate cancellation in REGAPIC with an RpcException
In both Grpc.Core and Grpc.Net.Client, for both streaming calls and unary calls, if a cancellation token is cancelled the result is an RpcException with a StatusCode of Cancelled.

We should do the same in REGAPIC.

### PR DESCRIPTION
**Pull request by jskeet**
Fixes #648

### CODE REVIEW

File: Google.Api.Gax.Grpc/Rest/RestChannel.cs
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/RestChannel.cs:71]: I think the linked token built by SendAsync should be passed to ReadResponseAsync as well and used to wrap the content reading call with something similar to AwaitWithCancellation.

**Review by amanda-tarafa — COMMENTED** [review]
I've left some comments, but I might be missing something.

File: Google.Api.Gax.Grpc/Rest/RestChannel.cs
Comment (jskeet) [reply Google.Api.Gax.Grpc/Rest/RestChannel.cs]: I think my change (new commit) should fix this.

File: Google.Api.Gax.Grpc/Rest/RestChannel.cs
Comment (amanda-tarafa) [reply Google.Api.Gax.Grpc/Rest/RestChannel.cs:71]: OK, but I still think that methodDisposalCts.Token should be passed to ReadResponseAsync (which currently expects no token), so the task that actually reads the content is cancallable as well.

File: Google.Api.Gax.Grpc/Rest/RestChannel.cs
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/RestChannel.cs:68]: nit: methodDisposalCts => callCancellationTokenSource (just so that is the same as in the streaming call)

**Review by amanda-tarafa — COMMENTED** [review]
One comment left, otherwise looks good.

File: Google.Api.Gax.Grpc/Rest/RestChannel.cs
Comment (jskeet) [reply Google.Api.Gax.Grpc/Rest/RestChannel.cs:68]: Done.

**Comment by jskeet** [comment]
> OK, but I still think that methodDisposalCts.Token should be passed to ReadResponseAsync (which currently expects no token), so the task that actually reads the content is cancallable as well.

I don't think that's required:

- ReadResponseAsync awaits the task it's passed, which is returned from SendAsync (which sets up cancellation)
- The SendAsync call uses HttpCompletionOption.ResponseContentRead, which means the returned task shouldn't return until the response is already present
- This means that ReadAsStringAsync in ReadResponseAsync shouldn't actually block

It's worth noting that none of the calls within ReadResponseAsync actually takes a cancellation token anyway.

I'm going to merge without that change, and if the above doesn't convince you, we can revisit it :)

**Comment by jskeet** [comment]
Looks like I can't merge without another review after all :)

**Review by amanda-tarafa — APPROVED** [review]
LGTM

Convinced: I hadn't noticed the HttpCompletionOption.ResponseContentRead.

**Pull request merged**
